qi-test/tests/compiler.rkt

```diff
@@ -0,0 +1,43 @@
+#lang racket/base
+
+(provide tests)
+
+(require (for-template qi/flow/core/compiler)
+         rackunit
+         rackunit/text-ui
+         (only-in math sqr))
+
+(define tests
+  (test-suite
+   "compiler tests"
+
+   (test-suite
+    "deforestation"
+    ;; (~>> values (filter odd?) (map sqr) values)
+    (check-equal? (syntax->datum
+                   (deforest-rewrite
+                     #'(thread values
+                               (#%partial-application
+                                ((#%host-expression filter)
+                                 (#%host-expression odd?)))
+                               (#%partial-application
+                                ((#%host-expression map)
+                                 (#%host-expression sqr)))
+                               values)))
+                  '(thread
+                    values
+                    (esc
+                     (λ (lst)
+                       ((cstream->list
+                         (inline-compose1
+                          (map-cstream-next
+                           sqr)
+                          (filter-cstream-next
+                           odd?)
+                          list->cstream-next))
+                        lst)))
+                    values)
+                  "deforestation in arbitrary positions"))))
```

Comment:
I wonder about the value of these tests. They are strongly coupled to the implementation and effectively lock in the expected results (unless we will change all compiler/optimization tests when compilation and optimization change?). What would failure tell us?

OTOH, I can imagine it's nice to know that optimizations are actually happening…

I suggest we look at or ask other compilers to find out if/how their programmers verify that the optimizing parts of the compiler are working. Some possible communities come to mind: Racket/Chez, Rust, GCC, etc.

Reply:
I think we'd like these tests to just validate individual rewrite rules, rather than ensure that a particular expression compiles a certain way. At the moment, `deforest-rewrite` seems like one such low level rewrite rule, although I acknowledge the test appears quite specific. This was my attempt at fulfilling the need that seemed to emerge in recent discussions, see especially [this discussion](https://github.com/drym-org/qi/wiki/Qi-Compiler-Sync-Sept-1-2023#better-ways-of-validating-optimizations) and [this one](https://github.com/drym-org/qi/wiki/Qi-Compiler-Sync-Sept-15-2023#improved-testing-and-benchmarking). In any case, I agree that reaching out to other communities doing this kind of thing could give us some good ideas here.

Reply:
Ah, right. I'd missed or forgotten that deforest-rewrite is specific to deforestation (and we're not invoking the entire optimizer)... is it specific to fusion? Or could there be other optimization rules in this macro? If so, I'd still be concerned about test churn. (Indeed, I still partly am, simply because it seems like testing the implementation rather than the specification [making churn likely if the implementation changes]. But what is the testable spec for an optimizing function? I think we've stumbled on another hard problem ;P)

Reply:
Yeah I mean, one answer is that the internal implementation could be considered to be static enough to be a formal specification 😜  But that's probably not the best answer, and I think you're right that the better answers would be hard and interesting. fwiw I don't think we will be adding any more patterns to this rewrite rule and it is specific to fusion.